americangut/diversity_analysis.py

```diff
@@ -0,0 +1,1729 @@
+        splits = np.vstack((splits, np.array(clean)))
+
+    return splits[1:], levels[:level]
+
+
+def get_ratio_heatmap(data, ref_pos=None, log=None):
+    """Calculates a ratio array
+
+    Parameters
+    ----------
+    data : array
+        The data to be plotted
+    ref_pos : int, optional
+        The column to be used as the denomenator. If nothing is supplied,
+        the ratio will be calculated compared to the arithmatic mean of
+        all values
+    log : float
+        The value to use to caluclate the log of the data. If no value is
+        specified, no log will be taken.
+
+    Returns
+    -------
+    ratio : array
+        A numpy array with the the ratio of the data values to the reference
+        group.
+
+    """
+    if ref_pos is None:
+        ref = data.mean(1)
+    else:
+        ref = data[:, ref_pos]
+
+    ref = ((ref*np.ones((1, data.shape[0]))) *
+           np.ones((data.shape[1], 1))).transpose()
+
+    if log is None:
+        ratio = data / ref
+    else:
+        ratio = np.log(data / ref) / np.log(log)
+
+    return ratio
+
+
+def heatmap(data, ax=None,  cmap='RdBu_r', clims=None, cbar_size=11, **kwargs):
+    """Wraps matplotlib's heatmap and formats the colorbar
+
+    Parameters
+    ----------
+    data : array
+        The data to be plotted in the heat map
+   ax : matplotlib axis, optional
+        The axis where data should be plotted. If none, a new axis instance
+        will be created.
+    xticklabels : list, optional
+        The labels to be used for the x-axis, to describe the groups of
+        data.
+    yticklabels : list, optional
+        The labels for each group on the y-axis.
+    cmap : str, optional
+        The colormap name to use for plotting.
+    clims : list, optional
+        The limits on the colormap
+    xfont_angle : float, optional
+        The angle in degrees for the x tick label text.
+    xfont_align : {'left', 'right', 'center'}, optional
+        The horizonal alignment of the x tick label text. For rotated text,
+        an alignment or 'right' is recommended.
+    yfont_angle : float, optional
+        The angle in degrees for the x tick label text.
+    yfont_align : {'left', 'right', 'center'}, optional
+        The horizonal alignment of the x tick label text. For rotated text,
+        an alignment or 'right' is recommended.
+    xfont_size : int, optional
+        Default is 12. The size of the x tick labels
+    yfont_size : int, optional
+        Default is 12. The size of the y tick labels
+    cbar_size : int, optional
+        Default is 11. The size of the labels on the colorbar.
+
+    Returns
+    -------
+    ax: matplotlib axis
+        The axis with the data plotted.
+    cbar : matplotlib colorbar
+        The colorbar instance on the plot
+
+    """
+    # Checks the shape of the data is sane
+    mat_shape = data.shape
+    if ('xticklabels' in kwargs and kwargs['xticklabels'] is not None and not
```

Comment:
```python
if kwargs.get('xticklabels', None) is not None and len(kwargs.get('xticklabels', [])) != mat_shape[0]:
    # ...
```

Note, "xticklabels" is described as optional, but the conditional (as it is written in the PR) requires it